# Creating and Manipulating NumPy Arrays

# Concept Session

## One-dimensional Arrays

In [5]:
import numpy as np

### Q-1) MegaCity Builders is a construction company. They seek the help of their data analyst to generate employee codes from 1 to 1,000 for their part time work force, as they are not interested in RDBMS solution.

In [7]:
employee_codes = np.arange(1, 1001)
employee_codes

array([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,   11,
         12,   13,   14,   15,   16,   17,   18,   19,   20,   21,   22,
         23,   24,   25,   26,   27,   28,   29,   30,   31,   32,   33,
         34,   35,   36,   37,   38,   39,   40,   41,   42,   43,   44,
         45,   46,   47,   48,   49,   50,   51,   52,   53,   54,   55,
         56,   57,   58,   59,   60,   61,   62,   63,   64,   65,   66,
         67,   68,   69,   70,   71,   72,   73,   74,   75,   76,   77,
         78,   79,   80,   81,   82,   83,   84,   85,   86,   87,   88,
         89,   90,   91,   92,   93,   94,   95,   96,   97,   98,   99,
        100,  101,  102,  103,  104,  105,  106,  107,  108,  109,  110,
        111,  112,  113,  114,  115,  116,  117,  118,  119,  120,  121,
        122,  123,  124,  125,  126,  127,  128,  129,  130,  131,  132,
        133,  134,  135,  136,  137,  138,  139,  140,  141,  142,  143,
        144,  145,  146,  147,  148,  149,  150,  1

### Q-2)  MegaCity Builders has a total of 1,000 part-time employees. The annual salaries of these employees lie between $10,000 and $20,000. The company wishes to find the average annual salary of the employees. 

In [9]:
salaries = np.random.uniform(10000, 20000, 1000) 
salaries

array([19825.93670013, 15784.99089254, 13993.5145571 , 17484.74110247,
       13324.17890151, 11568.84580194, 19301.853427  , 18386.34806831,
       14185.13464508, 19506.17812393, 18555.39278208, 19413.59183637,
       10317.77467143, 17137.30556761, 10915.04609386, 14084.49443241,
       15095.55994594, 15936.03387814, 16649.02258404, 17649.82053783,
       11982.30960508, 14977.64495375, 11785.48083277, 12606.31970431,
       11262.51125922, 12312.22503691, 19548.4860536 , 17380.77711813,
       16426.74432265, 14718.44482336, 19414.42395792, 12296.53467646,
       19051.05841021, 19393.51662576, 18196.61328396, 19282.43723823,
       12415.47439072, 15316.03348808, 12675.52386416, 14243.66671466,
       18298.82400412, 19126.1787907 , 17083.99001253, 15772.96503519,
       14987.81633177, 14903.88150949, 18233.58269286, 18529.73703184,
       11911.32290462, 17095.54630122, 16191.18260514, 11306.56929088,
       15928.23442869, 18981.88978281, 11842.90496992, 13747.6792273 ,
      

In [11]:
average_salary = np.mean(salaries)
average_salary

14964.836332535602

### Q-3) The HR wishes to know how many employees are getting salaries more than $15,000 per year. 


In [15]:
high_salary = np.sum(salaries>15000)
high_salary

484

### Q-4) The employees with the codes 5, 124, 200, 332, 420, 589, 665, 743, 887, and 901 want to resign. As per the policy, the company does not save the data of part-time employees once they resign.  How can you delete the record/data of these employees? Consider the array "dup" to contain a copy of the employee codes.


In [19]:
dup = employee_codes
employee_codes = np.delete(employee_codes, [4,123,199,331,419,588,664,742,886,900], axis =0)
employee_codes

array([   1,    2,    3,    4,    6,    7,    8,    9,   10,   11,   12,
         13,   14,   15,   16,   17,   18,   19,   20,   21,   22,   23,
         24,   25,   26,   27,   28,   29,   30,   31,   32,   33,   34,
         35,   36,   37,   38,   39,   40,   41,   42,   43,   44,   45,
         46,   47,   48,   49,   50,   51,   52,   53,   54,   55,   56,
         57,   58,   59,   60,   61,   62,   63,   64,   65,   66,   67,
         68,   69,   70,   71,   72,   73,   74,   75,   76,   77,   78,
         79,   80,   81,   82,   83,   84,   85,   86,   87,   88,   89,
         90,   91,   92,   93,   94,   95,   96,   97,   98,   99,  100,
        101,  102,  103,  104,  105,  106,  107,  108,  109,  110,  111,
        112,  113,  114,  115,  116,  117,  118,  119,  120,  121,  122,
        123,  125,  126,  127,  128,  129,  130,  131,  132,  133,  134,
        135,  136,  137,  138,  139,  140,  141,  142,  143,  144,  145,
        146,  147,  148,  149,  150,  151,  152,  1

### Q-5) The HR wants to hire 10 new employees, assign them the employee codes from 1001 to 1010, and add the data in the existing employee codes. How can you accomplish this task? Consider the array "empid" which originally contains the employee IDs. 


In [21]:
empid = employee_codes
new_employee_codes = np.arange(1001, 1011) 
new_employee_codes

array([1001, 1002, 1003, 1004, 1005, 1006, 1007, 1008, 1009, 1010])

In [25]:
empid = np.concatenate((empid, new_employee_codes))
empid

array([   1,    2,    3, ..., 1008, 1009, 1010])

### Q-6) The company wants to know and display the top 10 highest salaries of employees. The array 'sal' contains the salaries of all employees.


In [27]:
top_10_salaries = np.sort(salaries)[-10:][::-1]
top_10_salaries 

array([19992.17646625, 19981.45743712, 19978.60206169, 19967.31923766,
       19954.33375049, 19949.68966971, 19939.25043895, 19921.45804364,
       19919.82554265, 19908.06483431])

## Reading data from csv files using NumPy

### Q-7) How can you read data from a csv file that contains the column "Employee ID" with 20 values?

### Ensure that the csv file is present in the current working directory.

In [29]:
np.genfromtxt(r"C:\Users\mdaaq\Downloads\EMPLOYEE.csv", delimiter= ',')

array([[nan, nan],
       [ 1., 22.],
       [ 2., 27.],
       [ 3., 31.],
       [ 4., 29.],
       [ 5., 27.],
       [ 6., 26.],
       [ 7., 27.],
       [ 8., 23.],
       [ 9., 27.],
       [10., 32.],
       [11., 28.],
       [12., 27.],
       [13., 22.],
       [14., 27.],
       [15., 35.],
       [16., 33.],
       [17., 28.],
       [18., 22.],
       [19., 25.],
       [20., 27.],
       [21., 37.],
       [22., 27.],
       [23., 24.],
       [24., 27.],
       [25., 28.],
       [26., 26.],
       [27., 27.],
       [28., 27.],
       [29., 29.],
       [30., 49.],
       [31., 26.],
       [32., 32.],
       [33., 26.],
       [34., 24.],
       [35., 24.],
       [36., 23.],
       [37., 28.],
       [38., 28.],
       [39., 22.],
       [40., 26.],
       [41., 32.],
       [42., 35.],
       [43., 33.],
       [44., 31.],
       [45., 30.],
       [46., 36.],
       [47., 22.],
       [48., 23.],
       [49., 32.],
       [50., 34.]])

## Two-dimensional Arrays (Tabular Representation of Data)

### Q-1) A company wishes to allocate the employee codes from 101 to 105 along with the ages 25, 26, 27, 28, and 29 (in years) to its employees. How can you represent this data using arrays?

In [31]:
employee_ids = np.arange(101, 106)
ages = np.arange(25, 30)


### Columnar view of the data: To see the employee IDs along with the ages in a tabular orientation

In [33]:
employee_data = np.column_stack((employee_ids, ages))
employee_data

array([[101,  25],
       [102,  26],
       [103,  27],
       [104,  28],
       [105,  29]])

### Creating 2D arrays with random numbers

### Q-2) Accessing the elements of 2D arrays

#### 2.1: Display the ages of all the employees from the emp_data.

In [50]:
ages

array([25, 26, 27, 28, 29])

#### 2.2: Display all the employee codes.

In [58]:
employee_codes

array([101, 102, 103, 104, 105])

#### 2.3: Display the age of 4th employee.

In [64]:
age_of_4th_employee = employee_data[3, 1]
age_of_4th_employee

28

#### 2.4: Display the ages from 3rd employee onwards.

In [68]:
ages_from_3rd_employee = employee_data[2:, 1]
ages_from_3rd_employee

array([27, 28, 29])

#### 2.5: Display the employee codes from 2nd employee onwards.

In [70]:
employee_codes_from_2nd = employee_data[1:, 0]
employee_codes_from_2nd

array([102, 103, 104, 105])

#### 2.6: How can you display the records for 2nd, 3rd, and 4th employee?

In [72]:
records_2nd_3rd_4th = employee_data[1:4, :] 
records_2nd_3rd_4th 

array([[102,  26],
       [103,  27],
       [104,  28]])

#### 2.7: Find the age of the employee with the employee code 102.

In [76]:
employee_code = 102
row_index = np.where(employee_data[:, 0] == employee_code)[0][0]  
age_of_employee_102 = employee_data[row_index, 1] 
age_of_employee_102

26

### Q-3)  Updating 2D arrays:

#### 3.1: The employee with the employee code 103 wants to update the age to 31. How can you accomplish this task?

In [80]:
employee_code = 103
row_index = np.where(employee_data[:, 0] == employee_code)[0][0] 
row_index

2

In [82]:
employee_data[row_index, 1] = 31
employee_data

array([[101,  25],
       [102,  26],
       [103,  31],
       [104,  28],
       [105,  29]])

#### 3.2: The company wants to add the annual salaries of all the employees in the existing record. Add the annual salaries 35,000, 45,000, 40,000, 30,000, and 50,000 for the employees with the employee codes from 101 to 105, respectively.

In [86]:
salaries = [35000, 45000, 40000, 30000, 50000] 
employee_data_with_salaries = np.column_stack((employee_data, np.array(salaries)))
employee_data_with_salaries

array([[  101,    25, 35000],
       [  102,    26, 45000],
       [  103,    31, 40000],
       [  104,    28, 30000],
       [  105,    29, 50000]])

#### 3.3: The HR wants to update the salary of the employee with the employee code 102 to 43,500. Help the HR to accomplish this task.

In [88]:
employee_code = 102
new_salary = 43500
row_index = np.where(employee_data_with_salaries[:, 0] == employee_code)[0][0]
employee_data_with_salaries[row_index, 2] = new_salary
employee_data_with_salaries

array([[  101,    25, 35000],
       [  102,    26, 43500],
       [  103,    31, 40000],
       [  104,    28, 30000],
       [  105,    29, 50000]])

### Q-4) 2-D Arrays Manipulation:

The CEO of the organization has suggested to present all possible reports related to employees’ data to the HR team to understand the spread of salaries of the employees to take the decision for future recruitment and manage human resources expenses. How can you accomplish this task?


#### 4.1: Help Thomas, the financial analyst, wants to find the total annual salary of all employees.

In [92]:
total_annual_salary = np.sum(employee_data_with_salaries[:, 2]) 
total_annual_salary 

198500

#### 4.2: Help Thomas to calculate the average age of employees.

In [94]:
average_age = np.mean(employee_data_with_salaries[:, 1])
average_age 

27.8

#### 4.3: The company has hired a new employee who is 27 years old, with an annual salary of 37,000 and the employee code 106. Add the employee details in the existing record of employees.

In [102]:
new_employee = np.array([[106, 27, 37000]])
employee_data_with_salaries = np.vstack((employee_data_with_salaries, new_employee))
employee_data_with_salaries 

array([[  101,    25, 35000],
       [  102,    26, 43500],
       [  104,    28, 30000],
       [  105,    29, 50000],
       [  106,    27, 37000]])

#### 4.4: The employee with the code 103 wishes to resign from the current job position. The company needs to remove the employee's data and update the record. How can you accomplish this task?

In [108]:
employee_code_to_remove = 103
row_index = np.where(employee_data_with_salaries[:, 0] == employee_code_to_remove)
employee_data_with_salaries = np.delete(employee_data_with_salaries, row_index, axis=0)
employee_data_with_salaries 

array([[  101,    25, 35000],
       [  102,    26, 43500],
       [  104,    28, 30000],
       [  105,    29, 50000],
       [  106,    27, 37000]])

#### 4.5: The company has decided to give a bonus amount to all the employees who are getting the annual salary more than $35,000. For this, the company needs to know the number of such employees along with their salaries.

In [112]:
bonus_eligible_employees = emp_data_with_salaries[emp_data_with_salaries[:, 2] > 35000]
num_employees_with_bonus = len(bonus_eligible_employees)
num_employees_with_bonus 

4

In [114]:
bonus_eligible_employees

array([[  102,    26, 43500],
       [  103,    31, 40000],
       [  105,    29, 50000],
       [  106,    27, 37000]])

In [1]:
a = print('hello')
a

hello
